code/controller/actuator.cpp

```diff
@@ -1,10 +1,176 @@
+#include <thread>
+#include <chrono>
+
 #include "actuator.h"
+#include "../utility/logger.h"
+
+Actuator::Actuator(const QString& serial_port, const PortSettings& settings, QextSerialPort::QueryMode mode) {
+    m_serial_port = new QextSerialPort(serial_port, settings, mode);
+	m_invert_x = 1;
+	m_invert_y = 1;
+    //on successful connection
+    if (m_serial_port->lastError() == 0) {
+        Logger::log(serial_port.toStdString() + " successfully opened!", Logger::INFO);
+        m_serial_port->flush();
+		// Send a renumbering request for the devices
+		setDeviceNumber();
+    }
+    else {
+        Logger::log("ERROR: " + serial_port.toStdString() + " could not be opened! " + m_serial_port->errorString().toStdString(), Logger::ERROR);
+    }
+}
+
+Actuator::Actuator(const Actuator& other) {
+	m_x_device = other.m_x_device;
+	m_y_device = other.m_y_device;
+	m_invert_x = other.m_invert_x;
+	m_invert_y = other.m_invert_y;
+
+	m_serial_port = other.m_serial_port;
+}
+
+char* const Actuator::convertDataToBytes(long int data) {
+	if (data < 0) {
+		data = intPow(BYTE_RANGE, 4) + data;
+	}
+
+	char result[DATA_SIZE];
+	
+	for (int i = DATA_SIZE - 1; i >= 0; --i) {
+		int temp = intPow(BYTE_RANGE, i);
+		result[i] = data / temp;
+		data = data - temp * result[i];
+	}
+
+	return result;
+}
+
+void Actuator::setDeviceNumber() {
+	char instr[] = { 0, 2, 0, 0, 0, 0 };
+	
+	m_serial_port->write(instr);
+
+	m_x_device = 0;
+	m_y_device = 1;
+}
+
+int const Actuator::intPow(int x, int p) {
+	if (p == 0) return 1;
+	if (p == 1) return x;
+
+	int tmp = intPow(x, p / 2);
+	if (p % 2 == 0) return tmp * tmp;
+	else return x * tmp * tmp;
+}
 
-Actuator::Actuator(Mode)
-{
+int Actuator::setSerPort(const QString& serial_port) {
+	if (m_serial_port->portName() == serial_port) {
+		return 0;
+	}
+	//close the current port
+	if (m_serial_port->isOpen()) {
+		m_serial_port->flush();
+		m_serial_port->close();
+	}
+	
+	m_serial_port->setPortName(serial_port);
+	if (m_serial_port->lastError() != 0) {
+		Logger::log(m_serial_port->errorString().toStdString(), Logger::ERROR);
+		return -1;
+	}
+	return 0;
+}
+
+int Actuator::changeSettings(const PortSettings& settings) {
+	m_serial_port->setBaudRate(settings.BaudRate);
+	m_serial_port->setDataBits(settings.DataBits);
+	m_serial_port->setParity(settings.Parity);
+	m_serial_port->setStopBits(settings.StopBits);
+	m_serial_port->setFlowControl(settings.FlowControl);
+	m_serial_port->setTimeout(settings.Timeout_Millisec);
+
+	if (m_serial_port->lastError() != 0) {
+		Logger::log(m_serial_port->errorString().toStdString(), Logger::ERROR);
+		return -1;
+	}
+	return 0;
+}
 
+void Actuator::invertDevices() {
+	unsigned char temp = m_x_device;
+	m_x_device = m_y_device;
+	m_y_device = temp;
 }
 
 Actuator::~Actuator() {
+    delete m_serial_port;
+}
+
+void Actuator::move(Dir dir, int time) {

```

Comment:
the Dir version of the function is only a wrapper around vec2 for indexing from 1D commands, such as sending commands from the GUI and maybe keyboard arrow keys when they are implemented.